### PR TITLE
Definition based extration for custom extensions

### LINKED ISSUE #1381: Definition based extraction for Extensions
**Describe the bug**
 Definition based extraction for Extensions does not work since extraction is done using reflection. So in the example below from the [spec](http://build.fhir.org/ig/HL7/sdc/extraction.html#definition-extract) ":foo" will not exist in any object in HAPI.

> NOTE: Rather than specifying the id of an element that actually appears in a snapshot, it is also possible to 'extend' an id to walk further into the data types of the specified element. For example, a definition might be http://hl7.org/fhir/Patient#Patient.name.given even though the referenced profile's snapshot only contains Patient.name, not Patient.name.given. However, for more complex references (e.g. referring to a particular extension or a particular repetition), it will be necessary to define a distinct profile where slicing ensures that a single element id corresponds to the desired element. I.e. you would need to use http://somewhere.org/SomePatient#Patient.name.given:foo where 'foo' was a slice that referred to the first middle name in order to tie to a definition that specific.

**Component**
SDC library

**To Reproduce**

1. Use the [registration questionnaire](http://build.fhir.org/ig/WorldHealthOrganization/smart-anc/Questionnaire-ANCA.html) with the SDC
2. Extract resources from the Questionnaire Response
3. The Encounter will not contain the extension [Encounter.contactNumber](http://build.fhir.org/ig/WorldHealthOrganization/smart-anc/StructureDefinition-anc-encounter.html#Encounter.contactNumber)

**Expected behavior**
Expect the Encounter to contain the property contactNumber

cc: @jingtang10 @fredhersch

### PR DESCRIPTION
**Pull request by PallaviGanorkar**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1381

**Description**
if field is not present in resource then added it as definition based custom extension

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: Feature 

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 1]
```diff
@@ -448,6 +449,35 @@ object ResourceMapper {
...
+ * - resource's Base class instance See
+ * https://hapifhir.io/hapi-fhir/docs/model/profiles_and_extensions.html#extensions for more on
+ * custom extensions
+ */
+private fun addDefinitionBasedCustomExtension(
+  questionnaireItem: Questionnaire.QuestionnaireItemComponent,
+  questionnaireResponseItem: QuestionnaireResponse.QuestionnaireResponseItemComponent,
+  base: Base
+) {
+  if (base is DomainResource) {
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:474]: i think we need to handle the case to add extension to complex data types as well.

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 2]
```diff
@@ -427,7 +428,7 @@ object ResourceMapper {
     if (questionnaireResponseItem.answer.isEmpty()) return
 
     // Set the primitive type value if the field exists
-    val fieldName = getFieldNameByDefinition(questionnaireItem.definition)
+    val fieldName = getFieldNameByDefinition(questionnaireItem.definition) // 1111111111111
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:431]: clean up the comment

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 3]
```diff
@@ -448,6 +449,43 @@ object ResourceMapper {
     } catch (e: NoSuchMethodException) {
       // Do nothing
     }
+
+    if (base.javaClass.getFieldOrNull(fieldName) == null) {
+      // If field not found in resource class add definition based extension
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:454]:
```suggestion
      // If field not found in resource class, assume this is an extension
```

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 4]
```diff
@@ -448,6 +449,43 @@ object ResourceMapper {
...
+      addDefinitionBasedCustomExtension(questionnaireItem, questionnaireResponseItem, base)
+    }
+  }
+}
+
+/**
+ * Adds definition based custom extension for Resource.
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:461]:
```suggestion
 * Adds custom extension for Resource.
```

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 5]
```diff
@@ -512,7 +550,7 @@ private fun updateField(
     updateFieldWithAnswer(base, field, answersOfFieldType.first())
   } catch (e: NoSuchMethodException) {
     // some set methods expect a list of objects
-    updateListFieldWithAnswer(base, field, answersOfFieldType)
+    updateListFieldWithAnswer(base, field, answersOfFieldType) // given name updated here
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:553]: revert this change

File: datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt [hunk 6]
```diff
@@ -2376,6 +2377,483 @@ class ResourceMapperTest {
...
+      .isEqualTo(9)
+  }
+
+  @Test
+  fun `extract() should perform definition based extraction for custom extensions with valueCodeableConcept`():
+    Unit = runBlocking {
+    @Language("JSON")
+    val questionnaire =
+      """{
+  "resourceType": "Questionnaire",
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt:2453]: fix the indentation here

File: datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt [hunk 7]
```diff
@@ -2376,6 +2377,483 @@ class ResourceMapperTest {
...
+    assertThat(patient).isNotNull()
+    assertThat(
+        patient.getExtensionByUrl(
+            "http://hl7.org/fhir/StructureDefinition/Patient#Patient.extension:tribe"
+          )
+          .value
+      )
+      .isInstanceOf(Coding::class.java)
+    assertThat(
+        (patient.getExtensionByUrl(
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt:2698]: if you do a type cast `as` you don't need this.

File: datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt [hunk 8]
```diff
@@ -2376,6 +2377,483 @@ class ResourceMapperTest {
...
+      .isEqualTo("hausa")
+    assertThat(
+        (patient.getExtensionByUrl(
+              "http://hl7.org/fhir/StructureDefinition/Patient#Patient.extension:tribe"
+            )
+            .value as
+            Coding)
+          .display
+      )
+      .isEqualTo("Hausa")
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt:2714]: combine these two by using a local variable for the `Coding` first.

File: datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt [hunk 9]
```diff
@@ -2376,6 +2377,483 @@ class ResourceMapperTest {
...
+            "http://fhir.org/guides/who/anc-cds/StructureDefinition/anc-encounter#Encounter.contactNumber"
+          )
+          .valueAsPrimitive
+          .value
+      )
+      .isEqualTo(9)
+  }
+
+  @Test
+  fun `extract() should perform definition based extraction for custom extensions with valueCodeableConcept`():
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt:2448]: discussed: please make this test case minimum to focus on what we're trying to test. having a lot of information in the questionnaire is noisy and makes the test hard to debug.

File: datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt [hunk 10]
```diff
@@ -2376,6 +2377,483 @@ class ResourceMapperTest {
...
+            )
+            .value as
+            Coding)
+          .display
+      )
+      .isEqualTo("Hausa")
+  }
+
+  @Test
+  fun `extract() should perform definition based extraction for Complex DataType with custom extensions`():
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt:2718]: same comment as above - make this test case minimum.

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt (hunk 2 above)
Comment (PallaviGanorkar) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:431]: Done

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt (hunk 3 above)
Comment (PallaviGanorkar) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:454]: Done

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt (hunk 4 above)
Comment (PallaviGanorkar) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:461]: Done

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt (hunk 5 above)
Comment (PallaviGanorkar) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:553]: Done

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt (hunk 1 above)
Comment (PallaviGanorkar) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:482]: Done

File: datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt (hunk 6 above)
Comment (PallaviGanorkar) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt:2453]: Done

File: datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt (hunk 7 above)
Comment (PallaviGanorkar) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt:2698]: Done

File: datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt (hunk 8 above)
Comment (PallaviGanorkar) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt:2714]: Done

File: datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt (hunk 9 above)
Comment (PallaviGanorkar) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt:2448]: Done

File: datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt (hunk 10 above)
Comment (PallaviGanorkar) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt:2718]: Done

File: datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt [hunk 11]
```diff
@@ -2376,6 +2379,246 @@ class ResourceMapperTest {
...
+          ]
+        }
+      """.trimIndent()
+    val iParser: IParser = FhirContext.forR4().newJsonParser()
+    val questionnaireObj =
+      iParser.parseResource(Questionnaire::class.java, questionnaire) as Questionnaire
+    val temperatureQuestionnaireResponse =
+      iParser.parseResource(QuestionnaireResponse::class.java, response) as QuestionnaireResponse
+    val bundle = ResourceMapper.extract(questionnaireObj, temperatureQuestionnaireResponse)
+    val encounter = bundle.entry[0].resource as Encounter
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt:2437]:
```suggestion
    val encounter = bundle.entry.single().resource as Encounter
```

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 12]
```diff
@@ -448,6 +449,43 @@ object ResourceMapper {
     } catch (e: NoSuchMethodException) {
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:449]: should there be a return statement after L448?

**Review by jingtang10 — APPROVED** [review]

File: datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt (hunk 11 above)
Comment (PallaviGanorkar) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt:2437]: Done

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt (hunk 12 above)
Comment (PallaviGanorkar) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:449]: Done

**Pull request merged**